src/core/SampleDecoder.cpp

```diff
@@ -0,0 +1,169 @@
+#include "SampleDecoder.h"
+
+#include <QFile>
+#include <QFileInfo>
+#include <QString>
+#include <memory>
+#include <sndfile.h>
+
+#ifdef LMMS_HAVE_OGGVORBIS
+#include <vorbis/vorbisfile.h>
+#endif
+
+#include "AudioEngine.h"
+#include "DrumSynth.h"
+#include "Engine.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+std::vector<SampleDecoder::Decoder> SampleDecoder::s_decoders = {
+	&SampleDecoder::decodeSampleSF,
+#ifdef LMMS_HAVE_OGGVORBIS
+	&SampleDecoder::decodeSampleOggVorbis,
+#endif
+	&SampleDecoder::decodeSampleDS
+};
+
+auto SampleDecoder::supportedAudioTypes() -> const std::vector<AudioType>&
+{
+	static auto s_audioTypes = std::vector<AudioType>{};
+	if (!s_audioTypes.empty()) { return s_audioTypes; }
```

Comment:
A pattern I like to use for initialising static variables is to put the initialising code into an immediately-invoked lambda (or even extract it into a whole other function), whose result is assigned to the variable. Then the variable can be made const, and no manual checking is needed to see if it has been initialised yet. For example:
```cpp
auto SampleDecoder::supportedAudioTypes() -> const std::vector<AudioType>&
{
    static const auto audioTypes = [] {
      // ...
      return types;
    }();
    return audioTypes;
}
```
(Also, I don't think we use the `s_` prefix for local static variables, but there may not yet be a consensus.)

Reply:
> (Also, I don't think we use the s_ prefix for local static variables, but there may not yet be a consensus.)

clangd gives an annoying warning if its removed. I will just leave it for now. As I mentioned, we should probably just take another look at our `.clang-tidy` file.